[KMS] Creating the doc. (#295)
localstack/docs: content/en/aws/kms/index.md

---
title: "Key Management Service (KMS)"
linkTitle: "Key Management Service (KMS)"
categories: ["LocalStack Community"]
description: >
  Key Management Service (KMS)
---

Key Management Service (KMS) is mainly about
- creation and management of cryptographic keys

> - use of these keys to encrypt / decrypt data or to sign messages and to verify signatures for signed ones

Reviewer: Please use full sentences rather than points in the initial introduction. It would also be recommended to use camel case for writing sentences.

> AWS documentation on KMS amd all its operations can be found [here](https://docs.aws.amazon.com/kms/latest/APIReference/Welcome.html).

Reviewer 2: ```suggestion
AWS documentation on KMS and all its operations can be found [here](https://docs.aws.amazon.com/kms/latest/APIReference/Welcome.html).
```

> AWS documentation on KMS and all its operations can be found [here](https://docs.aws.amazon.com/kms/latest/APIReference/Welcome.html).

Reviewer: Avoid using `here` or `this page`, since we want the user to know where they are being directed to.

Keys in KMS roughly fall into the 3 categories:

> - Asymmetric encryption keys - holding a pair of a private and a public key. Asymmetric keys can be used either for data encryption and decryption with [Encrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Encrypt.html) / [Decrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html) or for creation and verification of digital signatures for messages with [Sign](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html) / [Verify](https://docs.aws.amazon.com/kms/latest/APIReference/API_Verify.html) operations. The same asymmetric key can't be used for both pairs of operations at the same time - when you create such a key, you have to specify in its [KeyUsage](https://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html#KMS-CreateKey-request-KeyUsage) parameter, which pair of operations the key will support.

Reviewer: ```suggestion
- Asymmetric encryption keys - These keys hold a pair of private and a public key. Asymmetric keys can be used either for data encryption and decryption with [Encrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Encrypt.html) / [Decrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html) or for creation and verification of digital signatures for messages with [Sign](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html) / [Verify](https://docs.aws.amazon.com/kms/latest/APIReference/API_Verify.html) operations. We can't use the same asymmetric key simultaneously for both pairs of operations. When you create such a key, you have to specify in its [KeyUsage](https://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html#KMS-CreateKey-request-KeyUsage) parameter which pair of operations the key will support.
```

> - Asymmetric encryption keys - holding a pair of a private and a public key. Asymmetric keys can be used either for data encryption and decryption with [Encrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Encrypt.html) / [Decrypt](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html) or for creation and verification of digital signatures for messages with [Sign](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html) / [Verify](https://docs.aws.amazon.com/kms/latest/APIReference/API_Verify.html) operations. The same asymmetric key can't be used for both pairs of operations at the same time - when you create such a key, you have to specify in its [KeyUsage](https://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html#KMS-CreateKey-request-KeyUsage) parameter, which pair of operations the key will support.

Reviewer: Please mention all of the categories in bold

> - Symmetric encryption keys - such keys consist of just one key. Symmetric keys can **not** be used to sign and to verify messages, are only used to encrypt and to decrypt data.

Reviewer: ```suggestion
- Symmetric encryption keys - These keys consist of just one key. Symmetric keys **cannot** be used to sign and verify messages. They are only used to encrypt and decrypt data.
```

> - Hash-Based Message Authentication Code (HMAC) keys, a special case of symmetric keys that are only used in creation and verification of message authentication codes (MACs) with [GenerateMac](https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateMac.html) / [VerifyMac](https://docs.aws.amazon.com/kms/latest/APIReference/API_VerifyMac.html) operations. You can read more about this type of keys [here](https://docs.aws.amazon.com/kms/latest/developerguide/hmac.html).

Reviewer: ```suggestion
- Hash-Based Message Authentication Code (HMAC) keys, a special case of symmetric keys that are only used in the creation and verification of message authentication codes (MACs) with [GenerateMac](https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateMac.html) / [VerifyMac](https://docs.aws.amazon.com/kms/latest/APIReference/API_VerifyMac.html) operations. You can read more about this type of key [on the HMAC keys documentation](https://docs.aws.amazon.com/kms/latest/developerguide/hmac.html).
```

> You can check [this page](https://docs.aws.amazon.com/kms/latest/developerguide/symm-asymm-compare.html) to see what other operations are supported for each type of keys. When you know what operations you require, this reference can help to figure out what key types are suitable for your goals.

Reviewer 2: tiny nits:
```suggestion
You can check [this page](https://docs.aws.amazon.com/kms/latest/developerguide/symm-asymm-compare.html) to see which other operations are supported for each type of key. Once you know which operations you require, this reference can help to figure out which key types are suitable for your goals.
```

> # Tutorial

Reviewer: Avoid using H1 tags here. All sub-sections should have H2 tags.

> # Tutorial

Reviewer: Please re-name this to `Getting started`. We would have a dedicated tutorial section coming up in the future.

Author: I will consider this, if you explain to me what would be the difference between "Getting started" and "Tutorial" sections.

Reviewer: A tutorial serves as a step-by-step walkthrough for folks who wish to accomplish a certain thing with LocalStack, usually have no idea of how to do it, and see a tutorial that exactly helps them with the given purpose. Tutorials are also intended to document complex workflows, like most of our Pro samples, which can touch on possible use cases.

Your content falls more in the scope of the "Getting started" guide, where we give enough basic information for someone to do something simplistic, in this case encrypting/decrypting something.

Author: This still sounds quite confusing to me. It seems that both tutorials and "getting started" guides are supposed to be step-by-step walkthroughs. The difference between them seems to be not that much in style, rather in scope / difficulty.

I believe that to our customers this is going to be a very confusing approach. Because the names of such sections - "Tutorial" / "Getting started" - do not provide enough information to understand the difference between them. Because the content is going to be very similar in style (step-by-step instructions). And because having multiple such set-by-steps on the same page would result in too big pages, that are going to be confusing and intimidating.

I would propose the following:
- We have just a single step-by-step set of instructions per service page. In some rare cases, where such instructions are like 2-3 commands, we might have 2 such step-by-steps.
- We create a separate section of documentation (separate from docs on services), where we collect and try to structure instructions on any complex scenarios our customers become interested in. We can discuss separately how we organize that.

Let's create a simple symmetric encryption key and use it to encrypt / decrypt something.

> A new key can be created in KMS with

Reviewer: Please use a colon

Author: Colons are used for clarifications. Would have been a proper place for a colon if this was something like "A new key can be created with the following command:". But this is not a clarification, the command is just basically a part of the sentence.

Reviewer: As you mentioned later, we can avoid splitting the sentence into two parts. In that case, "A new key can be created with the following command:" can be used as a suitable alternative, since I don't prefer having commands being a part of the same sentence.

Author: Ok. Changed the phrasing a bit in multiple places to have colons before code blocks.

> ```sh

Reviewer 2: nit: would be better to use the `<command>` macro here, and prefix the command with `$`. (there's a few examples in other docs..)

Author: Done.

$ awslocal kms create-key
{{</ command >}}

By default, this command creates a symmetric encryption key, so we do not even have to supply any additional arguments. In the output we want to pay attention to the ID of the newly created key, `010a4301-4205-4df8-ae52-4c2895d47326` in this case:

```sh
{
    "KeyMetadata": {
        "AWSAccountId": "000000000000",
        "KeyId": "010a4301-4205-4df8-ae52-4c2895d47326",
        "Arn": "arn:aws:kms:us-east-1:000000000000:key/010a4301-4205-4df8-ae52-4c2895d47326",
        "CreationDate": 1665432947.493433,
        "Enabled": true,
        "Description": "",
        "KeyUsage": "ENCRYPT_DECRYPT",
        "KeyState": "Enabled",
        "Origin": "AWS_KMS",
        "KeyManager": "CUSTOMER",
        "CustomerMasterKeySpec": "SYMMETRIC_DEFAULT",
        "KeySpec": "SYMMETRIC_DEFAULT",
        "EncryptionAlgorithms": [
            "SYMMETRIC_DEFAULT"
        ],
        "MultiRegion": false
    }
}
```

> If we lose the ID, we can always use

Reviewer: Please use a colon

Author: Same as above.

{{< command >}}
$ awslocal kms list-keys
{{</ command >}}

> to list IDs and [Amazon Resource Identifiers](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) (ARNs) of all the available keys. Then, if necessary, we can also check all the details about a key with a given key ID or ARN like this:

Reviewer: `to` —> `To`

> to list IDs and [Amazon Resource Identifiers](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) (ARNs) of all the available keys. Then, if necessary, we can also check all the details about a key with a given key ID or ARN like this:

Reviewer: `Then, if necessary` —> `If necessary`

{{< command >}}
$ awslocal kms describe-key --key-id 010a4301-4205-4df8-ae52-4c2895d47326
{{</ command >}}

> We can use the key now to encrypt something. Let's say, `some important stuff`:

Reviewer: Please remove backticks if you are not pointing to any variable, ID, code, or anything where you want the user to have a special attention. Using italics is fine.

{{< command >}}
$ awslocal kms encrypt \
      --key-id 010a4301-4205-4df8-ae52-4c2895d47326 \
      --plaintext "some important stuff" \
      --output text \
      --query CiphertextBlob \
  | base64 --decode > my_encrypted_data
{{</ command >}}

Why do we require such a complicated command? The `Encrypt` Operation itself looks like this:

{{< command >}}
$ awslocal kms encrypt \
      --key-id 010a4301-4205-4df8-ae52-4c2895d47326 \
      --plaintext "some important stuff"
{{</ command >}}

> And its output might be something like

Reviewer: Please use a colon

```sh
{
    "CiphertextBlob": "MDA4NDk4ZDYtMGU1ZS00NjU3LWEyZTQtMDliYTcwNDgyMmJkxw1eaLVUeUZA7bZzp+k7VAAAAAAAAAAAAAAAAAAAAAAetG9/5Znpw83/4xhwObc6Fc2B73y1ZvIigwahKopT0Q==",
    "KeyId": "arn:aws:kms:us-east-1:000000000000:key/010a4301-4205-4df8-ae52-4c2895d47326",
    "EncryptionAlgorithm": "SYMMETRIC_DEFAULT"
}
```

> The value of the `CiphertextBlob` field is a Base64-encoded result of encryption of our plaintext. In order to use KMS to decrypt this, we have to get rid of this Base64-encoding first. So to our `Encrypt` operation we add the `output` parameter, requesting the output to be in the text form instead of the default JSON. We also supply the `query` parameter, so the output only contains the value of the specified `CiphertextBlob` field. Then we feed the result to `base64` tool, asking it to remove the Base64 encoding, producing a binary output. We save that output to a file for future use. **Note:** Careful with the `CiphertextBlob` value for the `query` parameter - the AWS CLI doesn't check if the requested field is even if the output, so if you have a typo, the command would succeed, but the data sent to `base64` is going to be just a word `None`. Which is going to lead to unexpected results later. 

Reviewer: ```suggestion
The value of the `CiphertextBlob` field is a Base64-encoded result of the encryption of our plaintext. To use KMS to decrypt this, we must first get rid of this Base64-encoding. So to our `Encrypt` operation we add the `output` parameter, requesting the output to be in the text form instead of the default JSON. We also supply the `query` parameter, so the output only contains the value of the specified `CiphertextBlob` field. Then we feed the result to the `base64` tool, asking it to remove the Base64 encoding, producing a binary output. We save that output to a file for future use. **Note:** Careful with the `CiphertextBlob` value for the `query` parameter - the AWS CLI doesn't check if the requested field is even if the output, so if you have a typo, the command will succeed, but the data sent to `base64` is going to be just a word `None.` It will lead to unexpected results later. 
```

> When we want to, we can decrypt our file using the same KMS key:

Reviewer 2: ```suggestion
We can then decrypt the content of our file using the same KMS key:
```

{{< command >}}
$ awslocal kms decrypt \
      --ciphertext-blob fileb://my_encrypted_data \
      --output text \
      --query Plaintext
  | base64 --decode
{{</ command >}}

> Here we ask KMS to decrypt contents of the binary file created during our previous `Encrypt` call. We do not have to specify the `key-id` while decrypting our file - when encrypting something with a symmetric key, AWS includes the key ID into the encrypted data, so can figure out itself which key to use for decryption. With asymmetric keys the `key-id` has to be specified, though.

Reviewer: ```suggestion
Here we ask KMS to decrypt the contents of the binary file created during our previous `Encrypt` call. We do not have to specify the `key-id` while decrypting our file - when encrypting something with a symmetric key, AWS includes the key ID into the encrypted data so we can figure out which key to use for decryption. The `key-id` has to be specified with asymmetric keys.
```

Author: `so can figure out` here refers to AWS, not to users.
`With asymmetric keys` refers to the phrase `encrypting something with a symmetric key`. So the meaning is "with symmetric keys it is like this, with asymmetric it is different".

As with the previous `Encrypt` call, we have to get rid of the Base64-encoding to actually get our data, so we use the `output` and `query` parameters along with the `base64` tool here as well. If everything went fine, the output is going to be

```sh

> some important stuff

Reviewer: Please give atleast a sample output for the end user to verify if they followed the example correctly

Author: ```sh
some important stuff
```
**is** the output of the command above.
